scripts/anvil_tools/set_up_anvil_workspaces.py

```diff
@@ -0,0 +1,255 @@
+        print("Check output file for error details.")
+        return(response.json())
+
+
+def setup_workspaces(tsv, project="anvil-datastorage"):
+    """Get the workspace and associated auth domain from input tsv file."""
+
+    # read full tsv into dataframe
+    setup_info_df = pd.read_csv(tsv, sep="\t")
+
+    # create df for output tsv file
+    col_names = ["input_workspace_name", "input_auth_domain_name",
+                 "auth_domain_email", "auth_domain_creation_error",
+                 "workspace_link", "workspace_creation_error",
+                 "workspace_ACLs", "workspace_ACLs_error",
+                 "workspace_setup_status"]
+    output_df = pd.DataFrame(columns=col_names)
+
+    # per row in tsv/df
+    for row in setup_info_df.index:
+        # authorization domain
+        auth_domain_name = setup_info_df.loc[row].get(key='auth_domain_name')
+        output_df.at[row, "input_auth_domain_name"] = auth_domain_name
+        auth_status = create_auth_domain(auth_domain_name)
+
+        if auth_status != 201:  # if auth creation fail
+            output_df.at[row, "auth_domain_creation_error"] = auth_status  # write error response to col
+            continue  # go to next row in df
+        else:  # if auth creation success
+            # TODO: will i need auth_domain_email again?
+            auth_domain_email = f"{auth_domain_name}@firecloud.org"
+            output_df.at[row, "auth_domain_email"] = auth_domain_email
+
+        # workspace
+        workspace_name = setup_info_df.loc[row].get(key='workspace_name')
+        output_df.at[row, "input_workspace_name"] = workspace_name
+        create_ws_request = make_create_workspace_request(workspace_name, auth_domain_name, project)  # get json for createWorkspace
+
+        create_workspace_status = create_workspace(create_ws_request, workspace_name, project)  # createWorkspace
+
+        # created or user wants to continue with existing
+        if create_workspace_status in [201, "Y"]:
+            output_df.at[row, "workspace_link"] = f"https://app.terra.bio/#workspaces/{project}/{workspace_name}"
+            # add member ACLs to workspace
+            member_status_code, member_response = add_members_to_workspace(workspace_name, auth_domain_name, project)
+            if member_status_code == 200:
+                output_df.at[row, "workspace_ACLs"] = member_response  # add emails to df
+                output_df.at[row, "workspace_setup_status"] = "Success"  # denotes final workspace setup step
+            else:
+                output_df.at[row, "workspace_ACLs_error"] = member_response
+                output_df.at[row, "workspace_setup_status"] = "Failed"
+        # if "N" (user does not want to continue with existing workspace) or other error
+        else:
+            output_df.at[row, "workspace_creation_error"] = create_workspace_status  # write error response to col
```

Comment:
as it is, if the user said "N" to "do you want to proceed with this existing workspace", you'll just write "N" to this column. you might consider a more explanatory "error" message in this case (probably something you want to assign farther up)

Reply:
If a user selects "N", it returns a specific message saying that the user chose not to overwrite the existing workspace! Its the create_workspace() function now but it was earlier in check_workspace_exists()

Reply:
:facepalm: I totally missed that! it's a little confusing that "Y" stays "Y" but "N" gets converted to the message... but this is ultimately a good outcome, so cool!